platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,479 @@
+     *
+     * <b>This method must be always balanced with and called after {@link #pauseCompaction()}. If
+     * there are more / less calls to resume compactions than to pause, or if they are called in a
+     * wrong order, it will result in deadlocks.</b>
+     *
+     * @throws IOException If an I/O error occurs
+     */
+    public void resumeCompaction() throws IOException {
+        try {
+            if (compactionWasInProgress.getAndSet(false)) {
+                assert currentCompactionWriter.get() == null;
+                assert currentCompactionReader.get() == null;
+                startNewCompactionFile(compactionLevelInProgress.getAndSet(0));
+            }
+        } finally {
+            snapshotCompactionLock.release();
+        }
+    }
+
+    /**
+     * Compact (merge) all files that match the given filter.
+     *
+     * @param reportDurationMetricFunction function to report how long compaction took, in ms
+     * @param reportSavedSpaceMetricFunction function to report how much space was compacted, in Mb
+     * @throws IOException if there was a problem merging
+     * @throws InterruptedException if the merge thread was interupted
+     */
+    @SuppressWarnings("unchecked")
+    public void compact(
+            CASableLongIndex index,
+            @Nullable final BiConsumer<CompactionType, Long> reportDurationMetricFunction,
+            @Nullable final BiConsumer<CompactionType, Double> reportSavedSpaceMetricFunction)
+            throws IOException, InterruptedException {
+        Instant timestamp = Instant.now(clock);
+
+        final UnaryOperator<List<DataFileReader<?>>> filesToMergeFilter;
+        final CompactionType compactionLevel;
+        if (isTimeForFullMerge(timestamp)) {
+            lastFullMerge = timestamp;
+            /* Filter nothing during a full merge */
+            filesToMergeFilter = dataFileReaders -> dataFileReaders;
+            compactionLevel = CompactionType.FULL;
+            logger.info(MERKLE_DB.getMarker(), "[{}] Starting Large Merge", storeName);
+        } else if (isTimeForMediumMerge(timestamp)) {
+            lastMediumMerge = timestamp;
+            filesToMergeFilter = readersOfLevel(1, config.maxNumberOfFilesInMerge());
+            compactionLevel = CompactionType.MEDIUM;
+            logger.info(MERKLE_DB.getMarker(), "[{}] Starting Medium Merge", storeName);
+        } else {
+            filesToMergeFilter = readersOfLevel(0, config.maxNumberOfFilesInMerge());
+            compactionLevel = CompactionType.SMALL;
+            logger.info(MERKLE_DB.getMarker(), "[{}] Starting Small Merge", storeName);
+        }
+
+        final List<? extends DataFileReader<?>> allCompactableFiles = dataFileCollection.getAllCompletedFiles();
+        final List<DataFileReader<?>> filesToCompact =
+                filesToMergeFilter.apply((List<DataFileReader<?>>) allCompactableFiles);
+        if (filesToCompact == null) {
```

Comment:
I believe, the list can also be empty, not just null.

Reply:
Yes, but if it's empty it's not going to pass the next condition:
```
if (filesCount < config.minNumberOfFilesInMerge()) 
```